src/apps/weblib/ts-api/chart.ts

```diff
@@ -75,35 +81,55 @@ export class Chart implements ChartInterface {
 
 	start(): void {
 		const ctx = {
-			update: (force: boolean): void => this.updateFrame(force)
+			update: (force: boolean): void => this.updateAndRender(force)
 		}
 		this._plugins.hook(Hooks.start, ctx).default(() => {
-			this.updateFrame()
+			this.updateAndRender()
 		})
 	}
 
-	updateFrame(force: boolean = false): void {
+	private updateAndRender(force: boolean = false): void {
+		this._update()
+		this._render(force)
+	}
+
+	private _update(): void {
+		const ctx: UpdateContext = {
+			timeInMSecs: null
+		}
+		this._plugins.hook(Hooks.update, ctx).default((ctx) => {
+			if (ctx.timeInMSecs) {
+				this._cChart.update(ctx.timeInMSecs)
+			}
+		})
+	}
+
+	private _render(force: boolean): void {
+		const control = force ? RenderControlMode.forced : RenderControlMode.disabled
 		const ctx: RenderContext = {
 			renderer: null,
-			timeInMSecs: null,
-			enable: true,
-			force,
+			control: control,
+			changed: this._changed,
 			size: { x: 0, y: 0 }
 		}
 		this._plugins.hook(Hooks.render, ctx).default((ctx) => {
-			if (ctx.size.x >= 1 && ctx.size.y >= 1 && ctx.timeInMSecs !== null && ctx.renderer) {
-				const render = ctx.force || (ctx.enable && this._needsUpdate)
-				ctx.renderer.canvas = this._cCanvas
-				this._module.registerRenderer(this._cCanvas, ctx.renderer)
-				this._cChart.update(this._cCanvas, ctx.size.x, ctx.size.y, ctx.timeInMSecs, render)
-				this._module.unregisterRenderer(this._cCanvas)
-				this._needsUpdate = false
+			if (ctx.size.x >= 1 && ctx.size.y >= 1 && ctx.renderer) {
```

Comment:
negate it and make it an early return.

Reply:
what do you mean?

Reply:
current version:
```
if (...) {
  // do stuff
}
// nothing
```
expected:
```
if (!(...))
  return; // nothing
  
// do stuff
```

This will cause not-nested tabulators